verifier/verifier.go

```diff
@@ -518,6 +536,69 @@ func verifyAuthenticTimestamp(outcome *notation.VerificationOutcome) *notation.V
 	}
 }
 
+func verifyRevocation(outcome *notation.VerificationOutcome, client *http.Client, logger log.Logger) *notation.ValidationResult {
+	r, err := revocation.New(client)
+	if err != nil {
+		logger.Debug("error while calling revocation.New: %s", err.Error())
+		return &notation.ValidationResult{
+			Type:   trustpolicy.TypeRevocation,
+			Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+			Error:  errors.New("unable to check revocation status"),
+		}
+	}
+
+	signingTime := outcome.EnvelopeContent.SignerInfo.SignedAttributes.SigningTime
+	revocationResult := r.Validate(outcome.EnvelopeContent.SignerInfo.CertificateChain, signingTime)
+
+	result := &notation.ValidationResult{
+		Type:   trustpolicy.TypeRevocation,
+		Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+	}
+
+	currResult := revocation_base.ResultOK
+	storedCertSubject := ""
+	for i, certResult := range revocationResult {
+		// Log all non-nil errors as debug
+		for _, err := range certResult.ServerResults {
+			if err != nil {
+				logger.Debugf("error for certificate #%d in chain with subject %v: %v", (i + 1), outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String(), err)
+			}
+		}
+		if currResult != revocation_base.ResultRevoked {
+			switch certResult.Result {
+			case revocation_base.ResultOK, revocation_base.ResultNonRevokable:
+				// do not overwrite any result
+				continue
+			case revocation_base.ResultRevoked:
+				currResult = revocation_base.ResultRevoked
+				storedCertSubject = outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String()
+				// after this is set, the switch will be skipped to avoid
+				// overwriting the revoked result
+			default:
+				// revocation_base.ResultUnknown
+				// overwrite result of OK, but not Revoked
+				currResult = revocation_base.ResultUnknown
+				// keep subject closest to leaf
+				if storedCertSubject == "" {
+					storedCertSubject = outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String()
+				}
+			}
+		}
```

Comment:
This logic seems bit brittle (the need for comments to explain what this code is doing is an indicator). 

Since both `ResultRevoked` and `ResultUnknown` result in overall failure of revocation, I suggest doing something like below

```
	isErrored := false
	revocationStatus := revocation_base.ResultUnknown
	erroredCert := ""
	for i, certResult := range revocationResult {
		for _, err := range certResult.ServerResults {
			if err != nil {
				logger.Debugf("error for certificate #%d in chain with subject %v: %v", (i + 1), outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String(), err)

		if certResult.Result == revocation_base.ResultRevoked || certResult.Result ==  revocation_base.ResultUnknown {
				isErrored = true
				revocationStatus = certResult.Result
				erroredCert = outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String()
				break
			}
		if isErrored {
				break
			}
		}
```

The difference between the two approaches is that one gives precedence to `Revoked` error over `Unkown` error another doesn't. Given that both errors ultimately result in signature verification failing, I don't think we need to worry about giving precedence at all? What are you thoughts?

Reply:
The reason for comments is a result of combining two separate processes into one loop in order to save some cycles per Pritesh's recommendation.

I still believe that there should be precedence for Revoked over Unknown. If even one certificate is revoked, then the chain should be considered revoked. I think this is important for users to be aware of. Let's take an example where cert 3 is Unknown and cert 5 is Revoked. If we just return at the first failure (in this case an Unknown), they may think that it is good to try again later (or may go into debug logs to find the error that occurred). However, if we prioritize revoked, then they will see easily that there is a certificate in the chain that is revoked (which may change how a user interprets or acts upon the response).

Reply:
Consensus was that the Revoked over Unknown precedence makes sense. Updated to make logic more clear and concise.